File: shoring/flywheel/src/main/java/com/panopset/flywheel/NamedMap.java
package com.panopset.flywheel;

import java.io.File;
import java.io.IOException;
import java.io.StringWriter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.TreeSet;
import com.panopset.compat.Fileop;
import com.panopset.compat.Stringop;

/**
 * Map wrapper with a map name.
 *
 * @param <K>
 *          Key type.
 * @param <V>
 *          Value type.
 */
public final class NamedMap<K, V> {

  /**
   * Report unused keys. Numeric keys are left out because they are generally
   * reserved for lists.
   * 
   * @param file
   *          File to write report to.
   */
  public static void reportUnusedKeys(final String file) throws IOException {
    KEY_REPORTER.reportUnusedKeys(file);
  }

  private static final KeyReporter KEY_REPORTER = new KeyReporter();

  private final String name;

  private final Map<K, V> map;

  public Map<K, V> getMap() {
    return map;
  }

  public NamedMap(final String newName) {
    this(newName, Collections.synchronizedMap(new TreeMap<K, V>()));
  }

  public NamedMap(final String newName, final Map<K, V> newMap) {
    name = newName;
    map = newMap;
  }

  public void put(final K key, final V value) {
    KEY_REPORTER.reportDefinedKey(name, key.toString());
    map.put(key, value);
  }

  public V get(final K key) {
    KEY_REPORTER.reportUsedKey(name, key.toString());
    return map.get(key);
  }

  public int getSize() {
    return map == null ? 0 : map.size();
  }

  public KeyReporter getKeyReporter() {
    return KEY_REPORTER;
  }

  /**
   * Key Reporter is used to help see any un-used variables quickly.
   */
  public static final class KeyReporter {

    KeyReport getKeyReport(final String source) {
      KeyReport rtn = keyReports.get(source);
      if (rtn == null) {
        rtn = new KeyReport();
        keyReports.put(source, rtn);
      }
      return rtn;
    }

    void reportUsedKey(final String source, final String key) {
      KeyReport keyReport = getKeyReport(source);
      if (!keyReport.getUsedKeys().contains(key)) {
        keyReport.getUsedKeys().add(key);
      }
    }

    void reportDefinedKey(final String source, final String key) {
      KeyReport keyReport = getKeyReport(source);
      if (!keyReport.getDefinedKeys().contains(key)) {
        keyReport.getDefinedKeys().add(key);
      }
    }

    static final int MAXIMUM_NUMERICS = 100;

    void reportUnusedKeys(final String file) throws IOException {
      StringWriter sw = new StringWriter();
      for (String source : keyReports.keySet()) {
        StringWriter hdr = new StringWriter();
        hdr.append("*********" + source + ":");
        hdr.append(Stringop.getEol());
        KeyReport keyReport = getKeyReport(source);
        for (String s : keyReport.getUsedKeys()) {
          keyReport.getDefinedKeys().remove(s);
        }
        for (int i = 0; i < MAXIMUM_NUMERICS; i++) {
          keyReport.getDefinedKeys().remove("" + i);
        }
        keyReport.getDefinedKeys().remove(ReservedWords.FILE);
        keyReport.getDefinedKeys().remove(ReservedWords.SCRIPT);
        keyReport.getDefinedKeys().remove(ReservedWords.SPLITS);
        keyReport.getDefinedKeys().remove(ReservedWords.TARGET);
        keyReport.getDefinedKeys().remove(ReservedWords.TEMPLATE);
        keyReport.getDefinedKeys().remove(ReservedWords.TOKENS);
        boolean firstTime = true;
        for (String s : keyReport.getDefinedKeys()) {
          if (firstTime) {
            sw.append(hdr.toString());
            firstTime = false;
          }
          sw.append(s);
          sw.append(Stringop.getEol());
        }
      }
      Fileop.write(sw.toString(), new File(file));
    }

    private final Map<String, KeyReport> keyReports = Collections
        .synchronizedSortedMap(new TreeMap<String, KeyReport>());

    public static class KeyReport {
      private final List<String> usedKeys = new ArrayList<String>();

      private final Set<String> dfndKeys = Collections
          .synchronizedSortedSet(new TreeSet<String>());

      public Set<String> getDefinedKeys() {
        return dfndKeys;
      }

      public List<String> getUsedKeys() {
        return usedKeys;
      }
    }
  }
}

File: code/shoring/flywheel/src/main/kotlin/com/panopset/flywheel/NamedMap.kt
package com.panopset.flywheel

import com.panopset.compat.Stringop.getEol
import java.io.IOException
import java.io.StringWriter
import java.util.*

/**
 * Map wrapper with a map name.
 *
 * @param <K>
 * Key type.
 * @param <V>
 * Value type.
</V></K> */
class NamedMap<K, V> (
    private val name: String,
    val map: MutableMap<K, V> = Collections.synchronizedMap(HashMap())
) {
    fun put(key: K, value: V) {
        keyReporter.reportDefinedKey(name, key.toString())
        map[key] = value
    }

    operator fun get(key: K): V? {
        keyReporter.reportUsedKey(name, key.toString())
        return map[key]
    }

    val size: Int
        get() = map.size

    /**
     * Key Reporter is used to help see any un-used variables quickly.
     */
    class KeyReporter {
        fun getKeyReport(source: String): KeyReport {
            var rtn = keyReports[source]
            if (rtn == null) {
                rtn = KeyReport()
                keyReports[source] = rtn
            }
            return rtn
        }

        fun reportUsedKey(source: String, key: String) {
            val keyReport = getKeyReport(source)
            if (!keyReport.usedKeys.contains(key)) {
                keyReport.usedKeys.add(key)
            }
        }

        fun reportDefinedKey(source: String, key: String) {
            val keyReport = getKeyReport(source)
            if (!keyReport.definedKeys.contains(key)) {
                keyReport.definedKeys.add(key)
            }
        }

        fun reportUnusedKeys() {
            val sw = StringWriter()
            for (source in keyReports.keys) {
                val hdr = StringWriter()
                hdr.append("*********").append(source).append(":")
                hdr.append(getEol())
                val keyReport = getKeyReport(source)
                for (s in keyReport.usedKeys) {
                    keyReport.definedKeys.remove(s)
                }
                for (i in 0 until MAXIMUM_NUMERICS) {
                    keyReport.definedKeys.remove("" + i)
                }
                keyReport.definedKeys.remove(ReservedWords.FILE)
                keyReport.definedKeys.remove(ReservedWords.SCRIPT)
                keyReport.definedKeys.remove(ReservedWords.SPLITS)
                keyReport.definedKeys.remove(ReservedWords.TARGET)
                keyReport.definedKeys.remove(ReservedWords.TEMPLATE)
                keyReport.definedKeys.remove(ReservedWords.TOKENS)
                var firstTime = true
                for (s in keyReport.definedKeys) {
                    if (firstTime) {
                        sw.append(hdr.toString())
                        firstTime = false
                    }
                    sw.append(s)
                    sw.append(getEol())
                }
            }
            //Fileop.write(sw.toString(), new File(file));
        }

        private val keyReports: MutableMap<String, KeyReport> = Collections
            .synchronizedSortedMap(TreeMap())

        class KeyReport {
            val usedKeys: MutableList<String> = ArrayList()

            val definedKeys: MutableSet<String> = Collections
                .synchronizedSortedSet(TreeSet())
        }

        companion object {
            const val MAXIMUM_NUMERICS: Int = 100
        }
    }

    companion object {
        /**
         * Report unused keys. Numeric keys are left out because they are generally
         * reserved for lists.
         *
         * @param file
         * File to write report to.
         */
        fun reportUnusedKeys() {
            keyReporter.reportUnusedKeys()
        }

        val keyReporter: KeyReporter = KeyReporter()
    }
}
